### PR TITLE
Snorkel Python for Labeling Datasets Programmatically

### LINKED ISSUE #4189: [Machine Learning] Snorkel Python for Labeling Datasets Programmatically
#### Proposed title of article
[Machine Learning]Snorkel Python for Labeling Datasets Programmatically
#### Proposed article introduction

Snorkel is a system for programmatically building and managing training datasets without manual labeling. Rather than hand-labeling thousands of data points by hand, Snorkel programmatically labels massive amounts of training data using labeling functions—rules, heuristics, and other custom complex operators.

Snorkel Flow automatically learns the different labeling functions’ accuracies, denoises and integrates them, and stores versioned packages and training data in Data Manager. Unlike with hand-labeled data, you create training data in Snorkel Flow using code, so you can audit, modify, or serve it almost instantly. Snorkel Flow makes it easy to share resources and training data, with others on the team.

Snorkel requires users to write rules that can label some portion of data. These rules are high-level rules that are written with the help of domain experts. In Snorkel terms, these rules are called Labeling Functions. These labeling functions are applied to all unlabeled data. The transformed data is sent to the core function of Snorkel known as Label Model which estimates the individual labeling function accuracy and correlations amongst them to assign labels to unlabelled data. The snorkel also provides slicing functions which are used to observe label model performance on slices of data

Finally, provides access to popular open-source model libraries that you can train on CPU- or GPU-based computing infrastructure. Snorkel Flow makes it easy to tune your models with automated hyperparameter search.

Snorkel has the following functions.
- Labeling Function : for labeling unlabelled data
- Transformation Function: for data augmentation
- Slicing Function: for data subset selection for important dataset
#### Key takeaways 
1. What is Snorkel Python and how to install it.
2. Key features of data labeling.
3. How to implement Labeling Function, Transformation Function, and Slicing Function.
4. Labelling a sample dataset.

#### Article quality
My article covers a unique and original project to explain the concepts of Snorkel Python. It's a practical guide that follows all the stages of preparing our dataset before labeling it with the snorkel functions.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by charles721**
### Article checklist

> If you have not already, please go over our [Resources Page](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md) for more tips and tricks on how to improve your overall technical writing so reviews are swifter, increase the chance of a payout and provide more value to our readers.

> For first-time contributors and for more details on our submission guidelines, **see our [Contributing Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**.

#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is 600x400 pixels and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #4189 

> Finally, delete the article checklist notes in blockquotes and submit your PR. We look forward to reviewing your article.

### CODE REVIEW

File: content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md [hunk 1]
```diff
@@ -0,0 +1,395 @@
+---
+layout: engineering-education
+status: publish
+published: true
+url: /snorkel-python-for-labeling-datasets-programmatically/
+title: Snorkel Python for Labelling Datasets Programmatically
+description: 
+author:
```
Comment (lalith1403) [inline content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md:8]: Add the author name. I was unable to find your folder

File: content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+
+Snorkel is a Python library that is used for data labelling. It programmatically manages and builds training datasets without manual labelling. In machine learning, labels are the target or the output variable in a dataset. This is what the model is attempting to predict.
+
+Instead of humans labelling large datasets manually, Snorkel assigns labels to the extensive training data automatically. This is done using a set of user rules, labelling functions, and other in-built techniques.
+
+Snorkel requires users to come up with functions that contain explicit rules. We will use these rules to label the unlabeled data.
+
+In this tutorial, we will have an unlabeled dataset that contains a list of sentences. The list of sentences is made up of questions, while others are general statements.
+
+This tutorial aims to be able to label a sentence as either a question or not a question. If a sentence is a question, it is labelled `1`, and a non-question sentence(general statement) is labelled `-1`. All this will be done programmatically using Snorkel.
```
Comment (lalith1403) [inline content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md:26]: aims to label a sentence

File: content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+- [Conclusion](#conclusion)
+- [References](#references)
+
+### Prerequisites
+
+To follow along easily, the reader should:
+
+- Be familiar with [Python programming.](/engineering-education/python-projects-for-beginners/)
+- Know about [machine learning modelling.](engineering-education/house-price-prediction/)
+- Know how to use [Google colab notebooks.](https://colab.research.google.com)
```
Comment (lalith1403) [inline content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md:53]: Google Colab. It is a proper noun.

File: content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+If this rule is met, the sentence is labelled as `QUESTION = 1`. If it's not met, the sentence is labelled `ABSTAIN = -1`.
+
+```python
+@labeling_function()
+def lf_keyword_lookup(x):
+  keywords = ["why","what","when","who","where","how"]
+ return QUESTION if any(word in x.sentences.lower() for word in keywords) else ABSTAIN
+```
+
+We have defined our function as `def lf_keyword_lookup(x)` with the help of the `@labeling_function()` method.
```
Comment (lalith1403) [inline content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md:244]: The function name does not include def. Remove def when referring to `lf_keyword_lookup`. 

Secondly, @labeling_function is not a method. It is a decorator.

File: content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+To find `?` in a sentence, we use Python [regular expression](https://docs.python.org/3/library/re.html).
+
+It searches through the sentence patterns until it finds a `?`.
+
+For further reading on Python regular expression, read this [documentation](https://docs.python.org/3/library/re.html).
+
+```python
+import re
+@labeling_function()
+def lf_regex_contains_what(x):r
```
Comment (lalith1403) [inline content/articles/snorkel-python-for-labeling-datasets-programmatically/index.md:265]: Check this code once. There is an extra character

**Review by lalith1403 — CHANGES_REQUESTED** [review]
Excellent article. Loved the way it is written. A few minor changes and we can move onto the final round of review.

**Comment by charles721** [comment]
Hello @lalith1403, thanks for the review, I have made the changes

**Comment by ahmadmardeni1** [comment]
Great work gentlemen @charles721 @lalith1403
